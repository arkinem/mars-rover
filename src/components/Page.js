import React from "react";
import styled from "styled-components";
import Header from "./Header";
import { calculateRoversPaths } from "../helpers/paths";
import Background from "./Background";
import { colors, shadow } from "../helpers/style";
import { tryParseInput } from "../helpers/input";
import MessageBox from "./MessageBox";
import Button from "./Button";
import InputInfoTooltip from "./InputInfoTooltip";
import { parseOutput } from "../helpers/output";
import { GiPathDistance } from "react-icons/gi";
import PathsModal from "./PathsModal/PathsModal";

class Page extends React.Component {
  state = {
    error: null,
    inputText: tempText,
    rovers: null,
    plateau: null,
    pathsModalOpen: false,
  };

  onConfirm = () => {
    const { plateau, error, rovers } = tryParseInput(this.state.inputText);

    if (!error) {
      if (this.state.error) this.setState({ error: null });

      const roversCalculated = calculateRoversPaths(
        plateau.maxX,
        plateau.maxY,
        rovers
      );

      this.setState({ rovers: roversCalculated, plateau });
    } else {
      this.setState({ error });
    }
  };

  render() {
    return (
      <Container>
        <Background />
        <Content>
          <Header />
          <Label>Insert your input here:</Label>
          <TextInputContainer>
            <TextInput
              value={this.state.inputText}
              onChange={(e) => this.setState({ inputText: e.target.value })}
            />
            <InputInfoTooltip />
          </TextInputContainer>

          <MessageBox text={this.state.error} />
          <ConfirmButton label="Calculate" onClick={this.onConfirm} />
          <TextOutputContainer>
            <TextOutput
              readOnly={true}
              value={parseOutput(this.state.rovers, false)}
            />
            {this.state.rovers && this.state.plateau && (
              <>
                <DetailsModalButton
                  label={<GiPathDistance color="white" />}
                  onClick={() => this.setState({ pathsModalOpen: true })}
                />
                <PathsModal
                  plateau={this.state.plateau}
                  rovers={this.state.rovers}
                  isOpen={this.state.pathsModalOpen}
                  closeModal={() => this.setState({ pathsModalOpen: false })}
                />
              </>
            )}
          </TextOutputContainer>
        </Content>
      </Container>
    );
  }
}

export default Page;

const Container = styled.div`
  min-height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Content = styled.div`
  background-color: ${colors.surfaceDarker};
  padding: 16px;
  border-radius: 6px;
  min-width: 320px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  box-shadow: ${shadow.xl};
`;

const Label = styled.p`
  margin-bottom: 12px;
  font-weight: 300;
`;

const TextInputContainer = styled.div`
  position: relative;
  width: 100%;
`;

const TextOutputContainer = styled.div`
  position: relative;
  width: 100%;
`;

const TextInput = styled.textarea`
  width: 100%;
  resize: none;
  height: 150px;
`;

const TextOutput = styled(TextInput)`
  height: 150px;
`;

const ConfirmButton = styled(Button)`
  margin: 12px 0;
`;

const DetailsModalButton = styled(Button)`
  position: absolute;
  bottom: 15px;
  right: 10px;
  padding: 0;
  padding-top: 10px;
  font-size: 26px;
  width: 45px;
  height: 45px;
  border-radius: 100%;
`;

const tempText = `15 15
1 2 N
LMLMLMLMM
3 3 E
MMRMMRMRRM
14 14 S
MMMMRMMMMMMMMMLMMMMRMMMMMMRMMMMMMM
8 5 N
LMLMLMLMM
3 12 N
MMMMMMMMMMMRMMMMMM
7 7 N
LMLMLMLMM
9 9 N
LMLMLMLMM
`;
