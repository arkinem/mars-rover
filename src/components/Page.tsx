import React from "react";
import styled from "styled-components";
import { Header } from "./Header";
import Background from "./Background";
import { colors, shadow } from "../constants/style";
import { tryParseInput } from "../helpers/io";
import InputGroup from "./InputGroup";
import { Rover, calculateRoversPaths } from "../helpers/rover";
import { Plateau } from "../helpers/plateau";
import ResultsGroup from "./ResultsGroup";

type State = {
  inputText: string;
  error: string | null;
  rovers: Rover[];
  plateau: Plateau | null;
  pathsModalOpen: boolean;
};

export default class Page extends React.Component<{}, State> {
  state = {
    inputText: "",
    error: null,
    rovers: [],
    plateau: null,
    pathsModalOpen: false,
  };

  componentDidUpdate = (prevProps: any, prevState: State): void => {
    if (prevState.inputText !== this.state.inputText) {
      const { plateau, error, rovers } = tryParseInput(this.state.inputText);

      if (error && this.state.error !== error) {
        this.setState({ error });
      } else if (plateau) {
        this.setState({
          plateau,
          rovers: calculateRoversPaths(rovers, plateau),
        });
      }
    }
  };

  render() {
    const { rovers, plateau, error } = this.state;

    return (
      <Container>
        <Background />
        <Content>
          <Header />
          <InputGroup
            error={error}
            onConfirm={(inputText) => this.setState({ inputText })}
          />
          <ResultsGroup rovers={rovers} plateau={plateau} />
        </Content>
      </Container>
    );
  }
}

const Container = styled.div`
  min-height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Content = styled.div`
  background-color: ${colors.surfaceDarker};
  padding: 16px;
  border-radius: 6px;
  min-width: 320px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  box-shadow: ${shadow.xl};
`;
