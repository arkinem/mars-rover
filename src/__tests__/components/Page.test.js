import React from "react";
import { render } from "@testing-library/react";
import Page from "../../components/Page";

test("renders title", () => {
  const { getByText } = render(<Page />);
  const linkElement = getByText(/Mars Rover/i);
  expect(linkElement).toBeInTheDocument();
});
