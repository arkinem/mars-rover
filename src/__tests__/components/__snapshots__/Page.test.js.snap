// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page renders correctly 1`] = `
<div
  className="sc-fzoyAV eBXtFT"
>
  <div
    className="sc-AxirZ jbpFxm"
    id="tsparticles"
  >
    <canvas
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </div>
  <div
    className="sc-fzoLag gEXVFj"
  >
    <h1
      className="sc-AxjAm ekUCH"
    >
      Mars Rover
    </h1>
    <p
      className="sc-fzoXzr hiUqXh"
    >
      Insert your input here:
    </p>
    <div
      className="sc-fzpjYC gtCHiX"
    >
      <textarea
        className="sc-fznJRM dvQzZ"
        onChange={[Function]}
        value="15 15
1 2 N
LMLMLMLMM
3 3 E
MMRMMRMRRM
14 14 S
MMMMRMMMMMMMMMLMMMMRMMMMMMRMMMMMMM
8 5 N
LMLMLMLMM
3 12 N
MMMMMMMMMMMRMMMMMM
7 7 N
LMLMLMLMM
9 9 N
LMLMLMLMM
"
      />
      <div
        className="sc-AxhUy hHkuEX"
      >
        <svg
          className="sc-AxgMl fCKlsa"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
          />
        </svg>
        <div
          className="sc-AxheI eIelhe"
        >
          <p
            className="sc-Axmtr cdsaXg"
          >
            The 
            <span
              className="sc-AxmLO jCqoea"
            >
              first line of input
            </span>
             is the upper-right coordinates of the plateau, the lower-left coordinates are assumed to be 0,0.
          </p>
          <p
            className="sc-Axmtr cdsaXg"
          >
            The rest of the input is information pertaining to the rovers that have been deployed. Each rover has two lines of input. The first line gives the rover's position, and the second line is a series of instructions telling the rover how to explore the plateau.
          </p>
          <p
            className="sc-Axmtr cdsaXg"
          >
            The position is made up of two integers and a letter separated by spaces, corresponding to the x and y co-ordinates and the rover's orientation.
          </p>
          <p
            className="sc-Axmtr cdsaXg"
          >
            Example:
            

            5 5
            

            2 N LMLMLMLMM
            

            3 3 E MMRMMRMRRM
          </p>
          <p
            className="sc-Axmtr cdsaXg"
          >
            If you need more help, look here:
            

            <a
              className="sc-fzozJi bcrwuz"
              href="https://code.google.com/archive/p/marsrovertechchallenge/"
              target="_blank"
            >
              https://code.google.com/archive/p/marsrovertechchallenge/
            </a>
          </p>
        </div>
      </div>
    </div>
    <div
      className="sc-AxiKw byiTWX"
    />
    <button
      className="sc-AxhCb fTFSWH sc-fzoiQi kfmQsg"
      onClick={[Function]}
    >
      Calculate
    </button>
    <div
      className="sc-fznxsB fzbVxk"
    >
      <textarea
        className="sc-fznJRM sc-fznWqX hrLwwh"
        readOnly={true}
        value=""
      />
    </div>
  </div>
</div>
`;
